File: Algorithms/1260. Shift 2D Grid/kotlin/solution.kt
class Solution {
    fun shiftGrid(grid: Array<IntArray>, k: Int): List<List<Int>> {
        val newGrid = mutableListOf<MutableList<Int>>()
        val numRows = grid.size
        val numCols = grid[0].size
        val total = numRows * numCols
        val sub = k % total

        for (b in 0..numRows - 1) newGrid.add(mutableListOf<Int>())
        for (a in 0..total - 1) {
            val idx = (a - sub + total) % total
            newGrid.get(a / numCols).add(grid[idx / numCols][idx % numCols])
        }

        return newGrid
    }
}

File: Algorithms/1260. Shift 2D Grid/python/solution.py
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        num_rows, num_cols = len(grid), len(grid[0])
        new_grid = [[0] * num_cols for _ in range(num_rows)]
        for row in range(num_rows):
            for col in range(num_cols):
                new_col = (col + k) % num_cols
                new_row = (row + (col + k) // num_cols) % num_rows
                new_grid[new_row][new_col] = grid[row][col]
        return new_grid
